### ISSUE TITLE
Unable to get the access token using user name and password - 

### ISSUE THREAD

**Issue by seetharamireddy540** [AzureAD/azure-activedirectory-library-for-java#136]
Hi ,

I have been using the adl4j library to get the access token using user name and password. It was working on-until few weeks back and now it was broken.

I am getting an error as "client secret / client association " is required. I have also tried with lastest released jar which is  1.2.0 version and same error.

		<dependency>
			<groupId>com.microsoft.azure</groupId>
			<artifactId>adal4j</artifactId>
			<version>1.1.3</version>
		</dependency>

 public Future<AuthenticationResult> acquireToken(final String resource,
            final String clientId, final String username,
            final String password, final AuthenticationCallback callback) 

However, I don't find the method signature to provide the client secret along with user name and password ?

Can you pls help me to fix this issue?

Is it recommend to use this library for production code ?

Regards
Ram

**Comment by seetharamireddy540** [comment]
Here is the log trace 

com.microsoft.aad.adal4j.AuthenticationException: com.microsoft.aad.adal4j.AuthenticationException: {"error_description":"AADSTS70002: The request body must contain the following parameter: 'client_secret or client_assertion'.\r\nTrace ID: 708fffa5-eaeb-4722-98e1-ce75c9020d00\r\nCorrelation ID: 888b629b-136b-43b8-9403-90bd308db815\r\nTimestamp: 2017-04-04 22:30:41Z","error":"invalid_client"}
	at com.gehcit.auth.generator.TokenGeneratorImpl.getAccessTokenUsingUserNameAndPassword(TokenGeneratorImpl.java:124)
	at com.gehcit.auth.generator.TokenGeneratorTest.testTokenGenerationByUserNameAndPassword(TokenGeneratorTest.java:33)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:86)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:38)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:459)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:678)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:382)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:192)

**Comment by jviafora** [comment]
You don't need a client secret when you do user auth:

```
private static final String AUTH_RESOURCE = "https://graph.windows.net";
.
.
// AAD Config
String aadUrl = "https://login.windows.net/common";
String aadId = "9ba1a5c7-f17a-4de9-a1f1-6178c8d51223";
.
.
Future<AuthenticationResult> future = authContext.acquireToken(
        		AUTH_RESOURCE, aadId, username, password, null);
```

**Comment by jvandervelden** [comment]
We are seeing the same issue.

If I do a straight POST request to 'https://login.microsoftonline.com/common/oauth2/token' using SoapUI:

Without Client Secret:

![image](https://cloud.githubusercontent.com/assets/7331813/25189318/d229d304-24f6-11e7-9a77-f825a4c1be22.png)

With Client Secret:

![image](https://cloud.githubusercontent.com/assets/7331813/25189328/d91d0d84-24f6-11e7-8daa-b62c460776a7.png)

Looks like the client secret is now required and this library hasn't been updated.

**Comment by psignoret** [comment]
A client credential (`client_secret` or `client_assertion`) has always been required when the `client_id` in question refers to a confidential client (i.e. an app registered as a web app/web API).

However, the only scenario where the username/password flow is supported is for public clients (i.e. an app registered as a native client app), which are _not_ authenticated (and thus, don't have client credentials).

**Comment by praneetloke** [comment]
I ran into the same problem. I had to generate an access token for a client ID that is registered as a web app/API with a username/password. As @psignoret stated, I needed to pass the `client_secret` param along with the username/password but this library currently doesn't have a way to pass the `client_secret` when calling the `acquireToken(...)` with the username/password. I have opened #144 to hopefully get the library extended with a way to pass the additional param for cases when the client ID is a web app and not a native app.

**Comment by psignoret** [comment]
As I mentioned earlier, the only supported use of username/password is with apps registered as public clients (i.e. native client applications). You *must not* these these requests for confidential clients, as that is unsupported and the service will stop accepting those requests at any moment.

See also why #100 was rejected.

**Comment by MAHMOOD-TAHIR** [comment]
Hi @seetharamireddy540,
Did you find your solution? i'm also having the same problem

**Comment by jvandervelden** [comment]
What we ended up doing is making each native client it's own native application in Azure. We then issue a key to the native application from Azure which is the client secret. From there we can use the client secret authorization flow. We also have a double check in our application the verifies the client's application id to filter out rouge clients.

Since we started this, the native applications in Azure don't allow for keys to be generated. The flow still works if the client application is a webapi application in Azure. Maybe @psignoret can elaborate as to the reasoning behind this.

**Comment by psignoret** [comment]
A true native client application (public client, in OAuth 2.0 terminology) should *never* have any secret. The fact that before you were given the option of creating a secret in the UI was a confusing mistake, but it never made any sense whatsoever.

**Comment by sammycakes** [comment]
I got this same error. My problem was that even though the registered Native-Headless-Application Redirect URI type was public client (mobile and desktop), that wasn't enough for Azure to treat this as a native app. So from App registrations-->clicking on the Native-Headless-Application --> click on "Authentication", I had to scroll down to "Default Client Type" and select "Yes" to treat the application as a public client. Then, when I ran the app again, I did not get the error about the client secret.
